index.js

```diff
@@ -9,45 +8,54 @@ export class MaxBufferError extends Error {
 	}
 }
 
-export default async function getStream(inputStream, options = {}) {
-	if (!inputStream) {
-		throw new Error('Expected a stream');
-	}
-
-	const {maxBuffer = Number.POSITIVE_INFINITY, encoding = 'utf8'} = options;
-	const isBuffer = encoding === 'buffer';
+export async function getStreamAsBuffer(stream, options) {
+	return getStreamContents(stream, chunkTypes.buffer, options);
+}
 
-	const stream = new PassThroughStream({encoding: isBuffer ? undefined : encoding});
+export default async function getStream(stream, options) {
+	return getStreamContents(stream, chunkTypes.string, options);
+}
 
-	const newStream = compose(inputStream, stream);
+const getStreamContents = async (stream, {convertChunk, getContents}, {maxBuffer = Number.POSITIVE_INFINITY} = {}) => {
+	if (!stream) {
+		throw new Error('Expected a stream');
+	}
 
 	let length = 0;
 	const chunks = [];
 
-	const getBufferedValue = () => isBuffer ? Buffer.concat(chunks, length) : chunks.join('');
-
 	try {
-		for await (const chunk of newStream) {
-			chunks.push(chunk);
+		for await (const chunk of stream) {
+			const convertedChunk = convertChunk(chunk);
+			chunks.push(convertedChunk);
 			length += chunk.length;
 
 			if (length > maxBuffer) {
 				throw new MaxBufferError();
 			}
 		}
 
-		return getBufferedValue();
+		return getContents(chunks, length);
 	} catch (error) {
 		try {
-			error.bufferedData = getBufferedValue();
+			error.bufferedData = getContents(chunks, length);
 			// This throws when the buffered data is larger than the maximum length
 			// for a string or buffer
 		} catch {}
 
 		throw error;
 	}
-}
+};
 
-export async function getStreamAsBuffer(stream, options) {
-	return getStream(stream, {...options, encoding: 'buffer'});
-}
+const convertChunkToBuffer = chunk => Buffer.isBuffer(chunk) ? chunk : Buffer.from(chunk);
+
+const getContentsAsBuffer = (chunks, length) => Buffer.concat(chunks, length);
+
+const convertChunkToString = chunk => typeof chunk === 'string' ? chunk : chunk.toString();
+
+const getContentsAsString = chunks => chunks.join('');
+
+const chunkTypes = {
+	buffer: {convertChunk: convertChunkToBuffer, getContents: getContentsAsBuffer},
+	string: {convertChunk: convertChunkToString, getContents: getContentsAsString},
+};
```

Comment:
This allows adding more types, such as `ArrayBuffer`, which would be helpful to support `ReadableStream`.